### PR TITLE
chore: Modify build tool

### LINKED ISSUE #16: Replace build system with hatch
Right now we use `setuptools` and `setuptools_scm`
We need to move to `hatchling` and `hatch-semver`

### PR DESCRIPTION
**Pull request by fmotaf**
Closes #16

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -57,8 +60,11 @@ coverage = [
     {version = "^7.2", python = ">=3.7"}
 ]
 
-[tool.setuptools_scm]
-write_to = "mox/__version__.py"
+[tool.hatch.version]
+path = "hatch/__about__.py"
```
Comment (ivancrneto) [inline pyproject.toml]: I believe this should stay as `"mox/__version__.py"` and it will write automatically.

File: pyproject.toml [hunk 2]
```diff
@@ -43,7 +44,9 @@ keywords = [
 readme = "README.md"
 
 [tool.poetry.dependencies]
-setuptools_scm = { version = "^8.0", python = ">=3.8,<4" }
+hatch = { version = "^1.9.1", python = ">=3.9,<4" }
+hatchling = { version = "^1.21.0", python = ">=3.8,<4" }
+hatch-semver = { version = "^1.0.1",  python = ">=3.9,<4" }
```
Comment (ivancrneto) [inline pyproject.toml:50]: Excellent work! We need to drop Python 3.7 support before merging this, though. #37

**Review by ivancrneto — CHANGES_REQUESTED** [review]

File: pyproject.toml (hunk 1 above)
Comment (fmotaf) [reply pyproject.toml]: The thing is according to docs hatch looks for a variable named `__version__` inside the file pointed by `path` but in `mox/__version__.py` there's already a variable with a different value `'0.1.dev296+gb25cd38.d20240110'
` for the version which is (i believe) the version of the package and not the version of hatch, hence when i try to build using
` poetry run python -m hatch build` it fails with the message: 
`ValueError: Error getting the version from source 'regex': unable to parse the version from the file: mox/__version__.py
`
Docs: https://hatch.pypa.io/latest/version/#pyprojecttoml

**Comment by ivancrneto** [comment]
@fmotaf can you merge the latest main and solve the conflicts here? thank you!

File: pyproject.toml (hunk 1 above)
Comment (ivancrneto) [reply pyproject.toml]: From the docs, what should be in `__about__.py` is the project version, and not the version of hatch itself. The hatch version is already in `pyproject.toml`. Here is a suggestion so it works with the current `__version__.py` file:
```suggestion
path = "mox/__version__.py"
validate-bump = true
scheme = "semver"
pattern = "__version__ = version = \"(?P<version>[^\"]+)"
```

Comment (fmotaf) [reply pyproject.toml]: Cool, will patch it.

**Comment by ivancrneto** [comment]
Great work!

**Pull request merged**